JENKINS-74995 Add rootless support

lnlrbr opened the pull request:
Resolve problem described in [JENKINS-74995].
Summary: It's not possible to use rootless mode with the actual implementation as the plugin retrieves the user (agent) UID:GID via the 'id' command and the returned value is not the one expected in a rootless environment (i.e. "0:0"), so builds fail.
The provided change adds a test to look if we are in a rootless environment and returns the relevant value if needed.

### Testing done

Tests were done with builds executed on two Linux nodes, one configured with [Docker rootless mode](https://docs.docker.com/engine/security/rootless/) and the other with [Podman](https://podman.io/docs/installation). 
The pipelines include a 'docker' and a 'dockerfile' agent section.
There is some write access in the "build" stage to the mounted workspace.
If builds are done with the not modified plugin, the run part is done under the current (jenkins) user '-u 1001:1001', and builds can failed for not enough permissions on mounted volumes, or can lead to files owned to root inside the workspaces.
With builds done with the modified plugin, if a rootless environment is detected, the docker run part is done with the user parameter set to '-u 0:0'.


### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

jglick committed:
Clarifying password masking in `RegistryEndpointStepTest`

Commit:
Add rootless support

strangelookingnerd committed:
Use `jenkins.baseline` to reduce bom update mistakes (#328)

jonesbusy committed:
Require 2.452.4 and fix build (#327)

basil committed:
Revive test suite on Linux (#330)

MarkRx committed:
[JENKINS-75102] Fix Windows Docker running Windows container with spaces in workspace path (#326)

basil committed:
Revive test suite on Windows (#331)

Co-authored-by: Mark <MarkRx@users.noreply.github.com>

imoazeem commented:
+1

imoazeem commented:
Will this pull request be reviewed and merged to provide support for Docker rootless setup ?

imoazeem commented:
Does anyone have an update on whether this pull request will be reviewed and merged to add support for a rootless Docker setup ?

basil committed:
Migrate from EE 8 to EE 9 (#329)

dependabot[bot] committed:
Bump org.jenkins-ci.plugins:plugin from 5.5 to 5.7 (#334)

Bumps [org.jenkins-ci.plugins:plugin](https://github.com/jenkinsci/plugin-pom) from 5.5 to 5.7.
- [Release notes](https://github.com/jenkinsci/plugin-pom/releases)
- [Changelog](https://github.com/jenkinsci/plugin-pom/blob/master/CHANGELOG.md)
- [Commits](https://github.com/jenkinsci/plugin-pom/compare/plugin-5.5...plugin-5.7)

---
updated-dependencies:
- dependency-name: org.jenkins-ci.plugins:plugin
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

MarkEWaite committed:
Use `docker ps` to detect absence of docker permissions (#336)

The plugin BOM agents seem to have the `docker` command available but the
user running the agent is not authorized to use the `docker` command.
Previously that was detected by calling `docker ps` and detecting the
failure.

Restores a change made in

* https://github.com/jenkinsci/docker-workflow-plugin/pull/331

Testing done

Confirmed that I could see the same failure on a local computer as is seen on
https://ci.jenkins.io/job/Tools/job/bom/job/master/3968/testReport/org.jenkinsci.plugins.docker.workflow/DockerDSLTest/

The computer had Docker CE installed by the specific user running the
test did not have permission to access Docker.  Prior to this change,
the tests failed with the message:

CANNOT CONNECT TO THE DOCKER DAEMON AT UNIX:///VAR/RUN/DOCKER.SOCK. IS THE DOCKER DAEMON RUNNING?

After making this change, the tests pass on that computer with the
specific user that does not have permission to access Docker.

jglick committed:
Avoid excessive `Closure`s

jglick committed:
`void` return type was incorrect (but CPS-transformed code does not check) https://github.com/jenkinsci/docker-workflow-plugin/pull/337#discussion_r1970362720

jglick committed:
Skip `agentDockerWithCreds` unless both `docker.username` and `docker.password` are set

jglick committed:
Deleting `agentDockerWithCreds` test

jglick committed:
Compatibility for `DockerPipelineScript`

jglick committed:
Suppressing `DockerAgentTest.userHandbookDockerfile` in CI

jglick committed:
Dep released

jglick committed:
Make extension optional so it will be easier to delete the API later

dependabot[bot] committed:
Bump org.jenkins-ci.plugins:plugin from 5.7 to 5.9 (#344)

dependabot[bot] committed:
Bump io.jenkins.tools.bom:bom-2.479.x from 4228.v0a_71308d905b_ to 4488.v7fe26526366e (#345)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

concatime commented:
While replacing Docker with Podman in our CI, the Docker Workflow plugin would hang trying to start the first process inside the container. As you said, adding `-u 0:0` solved this issue, but I had some issues with deleting the workspace due to UID/GID mismatch. The solution that I found is to use `--userns keep-id` instead of `-u 0:0`. The `--userns` flag also plays well with `-u x:y` that Docker Workflow plugin passes to Docker by default.

dependabot[bot] committed:
Bump org.jenkins-ci.plugins:plugin from 5.9 to 5.12

Bumps [org.jenkins-ci.plugins:plugin](https://github.com/jenkinsci/plugin-pom) from 5.9 to 5.12.
- [Release notes](https://github.com/jenkinsci/plugin-pom/releases)
- [Changelog](https://github.com/jenkinsci/plugin-pom/blob/master/CHANGELOG.md)
- [Commits](https://github.com/jenkinsci/plugin-pom/compare/plugin-5.9...plugin-5.12)

---
updated-dependencies:
- dependency-name: org.jenkins-ci.plugins:plugin
  dependency-version: '5.12'
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

jglick committed:
[JENKINS-75679] Reproduced `FilePathPickle` usage in test

jglick committed:
More compelling test, using a containerized agent

jglick committed:
Fixed using `KeyMaterial2`

jglick committed:
`DockerDSLTest.firstDoNoHarm` flaky just like https://github.com/jenkinsci/credentials-binding-plugin/pull/279

jglick committed:
Dep released

sschuberth commented:
@jglick any chance to have a look?

dependabot[bot] committed:
Bump org.jenkins-ci.plugins:plugin from 5.12 to 5.18 (#358)

dependabot[bot] committed:
Bump org.testcontainers:testcontainers from 1.21.0 to 1.21.3 (#357)

dependabot[bot] committed:
Bump io.jenkins.tools.bom:bom-2.479.x from 4488.v7fe26526366e to 4969.v6ffa_18d90c9f (#356)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

Commit:
Add userns=keep-id option for podman
Remove UID:GID in rootless mode

lnlrbr committed:
Merge branch 'jenkinsci:master' into feature/rootless_mode

petermbauer commented:
We would also greatly appreciate if we could continue using this plugin in our company after switching to rootless Docker.

@lnlrbr some tests seem to be failing, is this an issue of the CI or is there something left to do?

@jglick i fail to understand your point. Almost anything can also be done with CLI commands but as we all know, the devil lies in the details, so getting it right is not as easy as it sounds in the first place. Apparently you were contributing to this plugin, but if this is not the case anymore, maybe @basil or @dwnusbaum can help out (i just tried to find some contributors based on merged PRs)?
Or is there something fundamentally wrong with the changes, or some reason for not supporting rootless setups?

petermbauer commented:
> > Almost anything can also be done with CLI commands but as we all know, the devil lies in the details, so getting it right is not as easy as it sounds in the first place.
> 
> Exactly. And a prepackaged version purporting to support everyone’s use cases perfectly is far harder.
> 

I would agree if this was about some additional niche feature, but adapting to an improvement of the underlying tool does not fall into this category.
However, it seems there is already a lot of negativity in the air here and you made up your mind, so it is what it is.

lnlrbr commented:
@petermbauer I've tried to rebase my fork from the base branch but for some reasons Github doesn't like it. At the same time, if proposals are not even considered and are treated with contempt, why bother?

pappnase99 commented:
For what it's worth,  I think the build issue in the 'linux-21' build is a result of the builder trying to start a container of the `maven:3-eclipse-temurin-17-alpine` image. However the alpine distro version of the maven image only supports the 'linux/amd64' arch. I suppose the proper image should be `3-eclipse-temurin-21` (without alpine) since this image also supports the 'linux/arm64/v8' arch, which seems to be the arch this builder is running on.
